### ISSUE TITLE
Bad OpenPGP implementation

### ISSUE THREAD

**Issue by the-djmaze** [the-djmaze/snappymail#89]
**Describe the bug**
The current OpenPGP implementation is incorrect.

1. Encrypt message should be with 'Recipient' public key (not your own key)
2. Sign a message should be with 'From' private key
3. Received signed "Content-Type: multipart/alternative" message can't be verified yet (only plain)
4. etc. etc.

Some limitations are caused by the current implementation in JavaScript.
Either the whole message body should be rendered as-is in JavaScript or handled in PHP with https://php.net/gnupg or others.

**Comment by ervee** [comment]
I remember a discussion back on RainLoop Git. If I recall correctly the keys (so also your private key!) are stored server side. In my case, I'm the server admin so I kind of trust this. But I would not trust anyone else with my private key. Even if I used my own server, I would not trust it if it was a VPS or some other shared service.

So if this is currently broken, I would consider this a good candidate to shrink the SnappyMail code :)

**Comment by the-djmaze** [comment]
> If I recall correctly the keys (so also your private key!) are stored server side.

No, private keys are in your browser localStorage (which also sucks).
An "encrypted" copy could be stored on the server (with symmetric encryption and passphrase) but that defeats the use of serverside gnupg.

RoundCube: uses gnupg
ProtonMail: does not use PGP (they say they do, but they send an e-mail with a link and you must open the link)
etc.

There are only 2 cases where we need private keys:
1. Sign a message
2. Decrypt received message

Public keys is no issue and we can still use for:
1. Verify received signature
2. Encrypt message to send

Solving the public keys issue would be the first step to solve.
This should be easy for webmail without security issues.

**Comment by GregThib** [comment]
> ProtonMail: does not use PGP (they say they do, but they send an e-mail with a link and you must open the link)
> etc.

Plaît-il ?
Some people send messages to me from protonmail, and I received them PGP-encrypted (as I have WKD configured in my domain, and then my public keys are easily discoverable).

Maybe you receive the message in this form when the user force encryption to a recipient who don't have a public key known or discoverable via WKD?
Cause, with protonmail we are able to encrypt to recipient whithout PGP-enabled encryption, with something like a password.

**Comment by GregThib** [comment]
Just my two cents: in the RainLoop issue feed, there is a lot of questions about PGP implementation, e.g. the use of WKD to discover keys, the opportunistic encryption, and a few things more.

Today, I think it's important to reach a core-base PGP encryption (encrypt, sign, verify) without getting too much pressure about the remaining (trust model, key discover, etc.).

As we can see with Thunderbird, since they have dropped the Enigmail support, their implementation is suffering too.

**Comment by the-djmaze** [comment]
@GregThib oh yes, i've added OpenPGP.js 5.0.1 for comparison and development.
Problem is that the library is not 100% compiled for web, it also (still) has Node.js code.
So i'm cleaning up the 1.5 MiB code to get more in line with the old 636 KiB
openpgp-sm.js

But still, i should look at GnuPG as well

**Comment by the-djmaze** [comment]
Working on this issue revealed that RainLoop never correctly verified signed messages.
A signature must be done on the whole `multipart/signed` but it never does.
It only checks if the pgp sgnature is valid.

**Comment by GregThib** [comment]
Oo
This is a pretty serious issue.

**Comment by the-djmaze** [comment]
> Oo This is a pretty serious issue.

Yep, but as you can see all the commits, there is work in progress :)

For now it uses GnuPG to verify PGP messages.
This method is tested and now works properly here.
1. uses [php pecl gnupg](https://www.php.net/gnupg)
2. store public keys in: `…/_data_/_default_/storage/[example.com]/[account]/.gnupg` (no UI yet)
3. JavaScript click on "🔒OpenPGP signed message (click to verify)" will call the new `MessagePgpVerify` (result not in UI yet)


Still todo regarding "verify":
1. Improve UI /#/settings/openpgp with public keys section for verifying received messages
2. Show verify results in UI
3. Alternatives to GnuPG extension: [Crypt_GPG](https://pear.php.net/package/Crypt_GPG/), [OpenPGP.js](https://openpgpjs.org/)

Todo after that:
1. Encrypt using public keys
2. Manage private keys
3. Decrypt messages with private keys
4. Sign messages with private keys

**Comment by the-djmaze** [comment]
Update: latest changes now properly load keyrings of:
* Mailvelope
* OpenPGP.js
* GnuPG

Composer window is revamped and has no PopupsComposeOpenPgp.
Instead the dropdown menu now has two options:
* Sign
* Encrypt

### Sign
Only enabled when chosen identity (from) has a private key.

### Encrypt
Only enabled when all recipients (to, cc, bcc) have a public key.
Also all recipients must be either in Mailvelope or OpenPGP.js or GnuPG and can't be mixed.
TODO: should be extended to lookup HKP servers and others to find the keys.


The system is still defunct and does not Sign nor Encrypt yet.
First i need to get all keyring systems to work properly.

**Comment by GregThib** [comment]
Impressive!

I wanted to congratulate you very much for your work on this tool. I am impressed by the speed and efficiency with which you deal with this problem, and thank you for the attention given to your users.

Gratefully!

**Comment by the-djmaze** [comment]
@GregThib thank you!

Working on this, i noticed many problems that have to be dealt with:
1. PHP PECL GnuPG doesn't import private keys. I need to call `gpg --import --homedir=/home/snappymail/__data__/,,,,/.gnupg secret_key.asc`
2. OpenPG.js and Mailvelope can't sign/encrypt attachments or the whole message (only cleartext)
3. etc.

So basically it is a PITA to sign/encrypt properly.
But hopefully i get there eventually.

Looking at the Enigma source of RoundCube it seems that it also should have issues with it.

**Comment by the-djmaze** [comment]
Dropped support for PEAR Crypt_GPG because it is missing features.
For best memory usage i want to stream data in/out the mailso imap stream with GnuPG in between.
That way it is easy to encrypt/decrypt/sign/verify large messages (20MiB+).

Crypt_GPG does have encryptFile but not an encryptStream.

So i'm building a new gpg.php class that shall solve it using gnupg and Crypt_GPG as examples.

**Comment by the-djmaze** [comment]
Improved detection of PGP/MIME encrypted messages.
It also nicely shows unencrypted attachments.

![afbeelding](https://user-images.githubusercontent.com/3752035/151013482-6484db9e-8d93-45bc-b947-130664d6d32a.png)

**Comment by the-djmaze** [comment]
Decryption now works using Mailvelope
![afbeelding](https://user-images.githubusercontent.com/3752035/151201767-a57b892f-c3e4-4497-bf4e-c7a1245ac440.png)

**Comment by the-djmaze** [comment]
I've update the demo at https://snappymail.eu/demo/
You can follow progress there and test using the new OpenPGP, GnuPG and [Mailvelope](https://mailvelope.com/en)

When using Mailvelope don't forget to edit the settings in Mailvelope -> Options -> Authorized Domains
![afbeelding](https://user-images.githubusercontent.com/3752035/151563698-e73f5061-73ba-4cd8-8543-171ff0eab3e6.png)

**Comment by the-djmaze** [comment]
Pfffft this was work: https://github.com/the-djmaze/snappymail/commit/e265a0f1c175a989e82ff2a6ce2411f9e9582a48

RainLoop modified the message HTML part in PHP on the server, and so did SnappyMail.
But while supporting PGP/MIME multipart encrypted messages, the issue with parsing HTML and attachments became a problem because that is server-side.

So i've moved all code to be client-side so that decrypted messages act the same.

**Comment by the-djmaze** [comment]
Demo at https://snappymail.eu/demo/ is updated with latest changes.

You can encrypt using Mailvelope

**Comment by the-djmaze** [comment]
Here's a preview release https://github.com/the-djmaze/snappymail/releases

**Comment by luluwebmaster** [comment]
Hello !

Note : My English is not very good, sorry for that.

I'm having a problem on my docker installation, based on @kouinkouin image ( #44 ).

Very simply, I can't decipher the encrypted emails.

Whether with OpenPGP or GnuPG, the decryption is done well (I have the message "Message encrypted by OpenPGP" after decryption), but the encrypted message is not replaced in the interface.

I note an interesting point, with GnuPG, when I look at the Ajax requests and I launch the decryption, the response contains the decrypted message, but the JSON does not seem correct ( See the screenshots ).

Note that I don't know if this is a problem specific to the Docker image, or if the problem also exists in the context of a classic installation.

Is this issue known / related to this issue?

Thanks in advance !

After click on "Decrypt" button :

![image](https://user-images.githubusercontent.com/15527061/218252415-c5d137ee-dc81-4781-8d4e-645de4af4f16.png)

Reponse of request when use GnuPG ( Response is decrypted, I can read message ) : 

![image](https://user-images.githubusercontent.com/15527061/218252685-7d307937-362f-4911-846d-30b8511f1868.png)

![image](https://user-images.githubusercontent.com/15527061/218252600-29fcee1a-0b92-4d1a-9404-9634b02e033e.png)

**Comment by the-djmaze** [comment]
Who encrypted the message?

Looking at it, it is quoted-printable encoded HTML and that is wrong.

There are 2 types of PGP encryption:
1. PGP/Inline which is plain text (not HTML)
2. PGP/MIME which can be plain and/or HTML with optional attachments and/or inline images

**Comment by luluwebmaster** [comment]
The message received has been encrypted by Thunderbird.

Could this be a problem with Thunderbird ?

**Comment by the-djmaze** [comment]
Correct!
https://superuser.com/questions/794959/what-is-introducing-quoted-printables-into-my-pgp-encrypted-emails

**Comment by the-djmaze** [comment]
If we make a workaround, how would other people with different applications see the broken Thunderbird?
- gmail
- kmail
- fairemail
- apple mail
- outlook
- microsoft mail
- etc. etc.

**Comment by the-djmaze** [comment]
Looking at it, the issue is different these days and i will look into it.

![afbeelding](https://user-images.githubusercontent.com/3752035/218703913-089ebe6f-2396-4d08-9070-34c81101889a.png)

**Comment by luluwebmaster** [comment]
Ok, not problem, I wait your feedback !

**Comment by luluwebmaster** [comment]
Hello,

After update of docker image ( 2.26.1 ), now that works correctly !

**Comment by LeifAndersen** [comment]
@the-djmaze FWIW, I seem to have the same problem that @luluwebmaster was describing. But the message was encrypted using mailvalope instead of thunderbird.
